### ISSUE TITLE
Bot consumption being met with 401 error

### ISSUE THREAD

**Issue by owenhey** [microsoft/BotBuilder-Samples#3395]
I have a root bot and another bot that is acting as a skill bot. I'm able to consume the skill bot in the composer using blank microsoft app ID and passwords in the emulator, but when I publish them both to Azure (with their associated app ID and passwords), the root bot runs into a 401 error (error invoking skill) when trying to consume the other bot.

I saw this (https://microsoft.github.io/botframework-solutions/help/known-issues/) thread about how this was a "known issue" and I made sure that the endpoint and app ID were correct in the app settings, so I am not sure what is happening.

Any help would be appreciated.

**Comment by stevkan** [comment]
Can you verify for me that you are supplying different app ID's for both the root and skill bots? Using the same app ID for both will generate that error.

**Comment by owenhey** [comment]
> Can you verify for me that you are supplying different app ID's for both the root and skill bots? Using the same app ID for both will generate that error.

Yes, the root bot has a different App ID and password than the skill bot. 

The root bot has the skill bot's App ID listed in the "BotFrameworkSkills" section of its appsettings.json, and the skill bot has a "*" for the "AllowedCallers" section of its appsettings.json.

**Comment by stevkan** [comment]
> I'm able to consume the skill bot in the composer using blank microsoft app ID and passwords in the emulator.

> The root bot has the skill bot's App ID listed in the "BotFrameworkSkills" section of its appsettings.json, and the skill bot has a "*" for the "AllowedCallers" section of its appsettings.json.

I'm a little confused by these two statements. In appsettings.json file of both the root bot and the skill, does each file have unique values entered for the `MicrosoftAppId` and `MicrosoftAppPassword` properties? And, to verify, the skill's app Id matches the app Id entered in the "BotFrameworkSkills" section of the root bot's appsettings.json file?

Also, will the skill bot run in Emulator if you have the app Id and password values entered in the appsettings.json file with matching values in Emulator? Or, does it only run when all properties are blank or only when the Emulator properties are blank?

**Comment by owenhey** [comment]
> > I'm able to consume the skill bot in the composer using blank microsoft app ID and passwords in the emulator.
> 
> > The root bot has the skill bot's App ID listed in the "BotFrameworkSkills" section of its appsettings.json, and the skill bot has a "*" for the "AllowedCallers" section of its appsettings.json.
> 
> I'm a little confused by these two statements. In appsettings.json file of both the root bot and the skill, does each file have unique values entered for the `MicrosoftAppId` and `MicrosoftAppPassword` properties? And, to verify, the skill's app Id matches the app Id entered in the "BotFrameworkSkills" section of the root bot's appsettings.json file?
> 
> Also, will the skill bot run in Emulator if you have the app Id and password values entered in the appsettings.json file with matching values in Emulator? Or, does it only run when all properties are blank or only when the Emulator properties are blank?

Yes - I apologize, I worded that a littles strange. The App ID and PW of the root bot and skill bot are completely unique (4 different values). 

In the BotFrameworkSkills section in the appsettings of the root bot, I have the App ID of the skillbot as well as its messaging endpoint (<mychatbotname>.azurewebsites.net/api/messages).

As for the emulator: As far as I can tell, the bot will never run in the emulator if I have values listed for the app ID and password. I've tried with test values like "a" and "b" as well as Microsoft-generated values (matching the values when launching it in the emulator, of course) with no success. I get an error 400 saying the microsoft app id or password is incorrect (I have made sure several times that they are copied correctly).

**Comment by stevkan** [comment]
> I'm able to consume the skill bot in the composer using blank microsoft app ID and passwords in the emulator, but when I publish them both to Azure...

@owenhey, thank you for your patience. Another clarifying question for you - are the root and skill bots based off of the samples available in this repo? While your various comments suggest 'yes', you also mention using Composer. Is the root bot a Composer bot or an SDK bot, like samples 80 and 81?

Assuming they are SDK bots, when you test them locally what is the messaging endpoints you are supplying to Emulator? And, be aware that the root bot's SkillHostEndpoint value in appsettings.json should be `http://localhost:3978/api/skills` and the SkillEndpoint within the BotframeworkSkills section should be `http://localhost:39783/api/messages`. However, when connecting in Emulator to either, the endpoints should both end with `/api/messages`.

In Azure, can you verify that the app Id and password for each is set in the bot's app service configuration?

Lastly, if none of that helps, would you be able to zip up your skill bot (removing any secrets, etc.) and post it here? Then, I could try to reproduce using your setup to see what is happening.

**Comment by owenhey** [comment]
> > I'm able to consume the skill bot in the composer using blank microsoft app ID and passwords in the emulator, but when I publish them both to Azure...
> 
> @owenhey, thank you for your patience. Another clarifying question for you - are the root and skill bots based off of the samples available in this repo? While your various comments suggest 'yes', you also mention using Composer. Is the root bot a Composer bot or an SDK bot, like samples 80 and 81?
> 
> Assuming they are SDK bots, when you test them locally what is the messaging endpoints you are supplying to Emulator? And, be aware that the root bot's SkillHostEndpoint value in appsettings.json should be `http://localhost:3978/api/skills` and the SkillEndpoint within the BotframeworkSkills section should be `http://localhost:39783/api/messages`. However, when connecting in Emulator to either, the endpoints should both end with `/api/messages`.
> 
> In Azure, can you verify that the app Id and password for each is set in the bot's app service configuration?
> 
> Lastly, if none of that helps, would you be able to zip up your skill bot (removing any secrets, etc.) and post it here? Then, I could try to reproduce using your setup to see what is happening.

I apologize, I misspoke when I said composer. I have been able to consume the bot in the _emulator_ locally. I'm not using composer at all (both bots are created using the SDK). 

The root bot is taken directly from sample 80, but the skill bot is one that I created myself (originally based off an echo bot I think but it has tons more functionality at this point). 

I am able to successfully consume the bot in composer. I run the root bot through `http://localhost:3978` and the skill bot through `http://localhost:39783`. Then, I supply `http://localhost:39783/api/messages` for the SkillEndpoint in the root bot. As mentioned before, I haven't been able to get the emulator to work when supplying app IDs and passwords, so both bots had empty strings for those sections in their `appsettings.json`. These settings allow consumption locally in the emulator.

However, when I publish both bots to Azure and test it in the webchat, it fails with the titular 401 error. I supply the correct app ID and passwords for both, and change the SkillEndpoint and other relevant fields in `appsettings.json` to reflect what's shown in Azure.

I have verified that both the app ID and password are set in the app services for each bot.

I would be able to zip up my skill bot, but I would prefer not to post in publicly here. Is there somewhere I can send it to just you?

**Comment by stevkan** [comment]
@owenhey, again, thank you for your patience. My profile on GH lists my email address. You are welcome to send it to me there. When you do so, please post a message here so I know to be on the look out for it.
